src/controllers/auth/resetPassword.ts

```diff
@@ -1,26 +1,108 @@
-  await check('password', 'Password must be at least 4 characters long').isLength({ min: 8 }).run(req)
-  await check('confirmPassword', 'Passwords do not match').equals(req.body.password).run(req)
+  // TODO: make this into a middleware where we can apply it on a route-by-route basis.
+  /** @see: https://expressjs.com/en/guide/using-middleware.html **/
+  if (!req.user) {
+    return res.status(401).json({ error: 'not logged in.' })
+  }
+  await check('currentPassword', 'Current Password is required.').exists().notEmpty().run(req)
+  await check('currentPassword')
+    .custom(async (value, { req }) => {
+      const validPassword = await req.user.comparePassword(value)
+
+      if (!validPassword) {
+        throw new Error('Current Password must match the saved password.')
+      }
+
+      return true
+    })
+    .run(req)
+  await check('password', 'Password must be at least 8 characters long.').isLength({ min: 8 }).run(req)
+  await check('passwordConfirmation', 'Passwords do not match.').equals(req.body.password).run(req)
 
   const errors = validationResult(req)
 
   if (!errors.isEmpty()) {
     return res.status(422).json({ errors: errors.array() })
   }
-  return res.status(501).send()
+
+  try {
+    await User.update(req.user, { password: await hash(req.body.password, 10) })
+    return res.status(204).send()
+  } catch (error) {
+    logger.error(`Unable to update password for user with id: ${req.user.id}`)
+    return res.status(400).json({ error: 'User not found' })
+  }
 }
 
-export const postReset = async (req: Request, res: Response): Promise<Response> => {
-  await check('password', 'Password must be at least 4 characters long.').isLength({ min: 8 }).run(req)
+// Guest reset password
+// /password/reset
+export const postGuestResetPassword = async (req: Request, res: Response): Promise<Response> => {
+  await check('password', 'Password must be at least 8 characters long.').isLength({ min: 8 }).run(req)
+  await check('userId', 'The userId must be an int.').isInt().run(req)
   await check('passwordConfirmation', 'Passwords must match.').equals(req.body.password).run(req)
 
   const errors = validationResult(req)
 
   if (!errors.isEmpty()) {
     return res.status(422).json({ errors: errors.array() })
   }
-  return res.status(501).json({ error: 'not implemented' })
+
+  const token = req.body.token
+  const userId = req.body.userId
+  let user: User
+
+  try {
+    user = await User.findOneOrFail(userId)
+  } catch (error) {
+    logger.error(error)
+    return res.status(404).json({ error: 'User not found for the given userId' })
+  }
+
+  // Find passwordReset record
+  let passwordReset: PasswordReset
+
+  try {
+    passwordReset = await PasswordReset.findOneOrFail({ where: { user } })
+  } catch (error) {
+    logger.error(error)
+    return res.status(404).json({ error: 'Token not found for the given email address' })
+  }
+
+  const createdAt = passwordReset.createdAt.toString()
```

Comment:
@larzhole Why not just use `DateTime.fromISO()` here since the date is likely something like `2020-06-30T20:02:24.000Z`? See more about parsing with luxon [here](https://moment.github.io/luxon/docs/manual/parsing.html).

Does this work?
```suggestion
  const createdAt = DateTime.fromISO(passwordReset.createdAt)
```

Reply:
@ericdowell No that's not working for me, it's saying it can't be parsed. There's gotta be a cleaner solution, but this was the best I could figure out after a few hours haha

```
invalid: Invalid
explanation: "the input "Fri Jul 10 2020 09:07:36 GMT-0500 (Central Daylight Time)" can't be parsed as ISO 8601"
reason: "unparsable"
```

Reply:
Seems createdAt is a Date object, which Luxon has DateTime.fromJSDate()

Otherwise you can use the Date.getTime() method on the createdAt Date object to get the milliseconds which can be passed to a Luxon DateTime method.

Reply:
I can confirm that `createdAt` on `User` is a `Date` object via logging the following:
```ts
import { logger } from '../logger'

class User {
  toJSON(): any {
    logger.info(`createdAt ${this.createdAt.constructor.name}, ${this.createdAt.getTime()}`)
//...
  }
}
```

We can update the types based on this. See commit: https://github.com/ericdowell/rerun/pull/11/commits/ebd2295d4f89afcf35d90b9d15cfa1f3f2153e98

Reply:
I updated this now to be much cleaner using your suggestion